### PR TITLE
ci: support image build on release

### LINKED ISSUE #32: CI tag version on tagged release
### What's Your Idea?

Currently the push on main causes both tests and a push of an "expected" version to the image repo.

I suggest that instead we should push the labeled tag on release publish rather than on push to main.

### Why Would This Be Helpful?

Currently, it's unclear if a tagged version is part of a WIP on main or an actual new version.

### Anything Else? (Optional)

_No response_

### Want to Help Out?

Yes! I'd love to help implement this

### Have You Considered Any Alternatives? (Optional)

_No response_

### Before Submitting

- [x] I've searched existing issues and confirmed this feature hasn't been requested yet

### PR DESCRIPTION
**Pull request by imnotjames**
## 📝 Description

CI currently does not build images as expected and builds a "versioned" tagged image on every push to `main` which is confusing.

NOTE: As soon as this is merged, pre-release docker images will start getting pushed - be ready for that.

**Linked Issue:** Fixes #32 , #45

## 🏷️ Type of Change

- [ ] Bug fix
- [ ] New feature
- [x] Enhancement to existing feature
- [ ] Refactor (no behavior change)
- [ ] Breaking change (existing functionality affected)
- [ ] Documentation update

### Updates the `build.gradle` to allow specifying version

To ensure the version gets in correctly, this updates the backend `build.gradle` to allow a property to be used for the version which can be passed via the command line.

This means the jar file will be `booklore-api-${VERSION}.jar`

### Drops labels from dockerfiles

The hard coded labels can be injected at runtime via the docker-metadata action so these were redundant.

### Creates a `publish.yml` workflow

This workflow publishes docker images and the jar file.

On `main` commits, the Docker image tagged in the ghcr.io repo as `main` and the commit hash.  The version is `0.0.0-SNAPSHOT` to denote this as not being an expected version.

On releases, the Docker image will be tagged as the release name, as well as `latest`.  The version is updated to be the semver on a tag name, so tagging as `vX.Y.Z` is encouraged.

The tagging definition is as such:

* `type=schedule` - will tag the image as `nightly` if a cron is set up to do nightly builds
* `type=semver,pattern={{version}}` - will tag the image on semver based on the tag name's "version" - eg, `v1.2.3` will create the tag `1.2.3`
* `type=ref,event=branch` - will tag the image on the branch name - so `main`
* `type=sha` - will tag the image on the short sha commit

One change is that the docker labels that were hardcoded have been removed and the docker metadata action has been included instead.  This sets the labels & tags automatically based on the Github repo information.

### Drops `build-and-push` from other pipelines

The other pipelines build-and-push are not publishing and `publish.yml` replaces them.

## 🧪 Testing (MANDATORY)

> **PRs without this section filled out will be closed.** "Tests pass" or "Tested locally" is not sufficient. You must provide specifics.

**Manual testing steps you performed:**

1. Applied to the `main` branch of a fork
2. Observed artifacts attached to run
3. Observed image built
4. Created a release in that fork
5. Observed artifacts attached to run
6. Observed image built
7. Observed API jar / UI tar attached to release.

Example Release: https://github.com/imnotjames/grimmory/releases/tag/v2.6.1
Example Action Run: https://github.com/imnotjames/grimmory/actions/runs/23250676755
Example Docker Image: docker pull ghcr.io/imnotjames/grimmory:2.6.1

**Regression testing:**

- The image building wasn't working before so no regressions there.

**Edge cases covered:**

-

**Test output:**

N/A

## 📸 Screen Recording / Screenshots (MANDATORY)

> Every PR must include a **screen recording or screenshots** showing the change working end-to-end in a running local instance (both backend and frontend). This means you must have actually built, run, and tested the code yourself. PRs without visual proof will be closed without review.

[Artifacts are attached to release automatically](https://github.com/imnotjames/grimmory/releases/tag/v2.6.1):

image

Docker Image has labels applied by metadata:

image

`main` and release name tags:

image

---

## ✅ Pre-Submission Checklist

> **All boxes must be checked before requesting review.** Incomplete PRs will be closed without review. No exceptions.

- [x] This PR is linked to an approved issue
- [x] Code follows project [backend and frontend conventions](../CONTRIBUTING.md#backend-conventions)
- [x] Branch is up to date with `develop` (merge conflicts resolved)
- [x] I ran the full stack locally (backend + frontend + database) and verified the change works
- [x] Automated tests added or updated to cover changes (backend **and** frontend)
- [x] All tests pass locally and output is pasted above
- [x] Screen recording or screenshots are attached above proving the change works
- [x] PR is a single focused change (one bug fix OR one feature, not multiple unrelated changes)
- [x] PR is reasonably scoped (PRs over 1000+ changed lines will be closed, split into smaller PRs)
- [x] No unsolicited refactors, cleanups, or "improvements" are bundled in
- [x] Flyway migration versioning is correct _(if schema was modified)_
- [ ] Documentation PR submitted to [booklore-docs](https://github.com/booklore-app/booklore-docs) _(if user-facing changes)_

### 🤖 AI-Assisted Contributions

> **If any part of this PR was generated or assisted by AI tools (Copilot, Claude, ChatGPT, etc.), all items below are mandatory.** You are fully responsible for every line you submit. "The AI wrote it" is not an excuse, and AI-generated PRs that clearly haven't been reviewed are the #1 reason PRs get closed.

No AI used.

---

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -49,22 +49,6 @@ RUN --mount=type=cache,target=/home/gradle/.gradle \
...
-LABEL org.opencontainers.image.title="WT-BookLore" \
-      org.opencontainers.image.description="WT-BookLore: A self-hosted, multi-user digital library with smart shelves, auto metadata, Kobo & KOReader sync, BookDrop imports, OPDS support, and a built-in reader for EPUB, PDF, and comics. This fork includes native support for the KOReader Sync plugin as well as all changes done by me." \
-      org.opencontainers.image.source="https://gitlab.worldteacher.dev/wt-booklore/wt-booklore" \
-      org.opencontainers.image.url="https://gitlab.worldteacher.dev/wt-booklore/wt-booklore" \
-      org.opencontainers.image.documentation="https://booklore.org/docs/getting-started" \
-      org.opencontainers.image.version=$APP_VERSION \
-      org.opencontainers.image.revision=$APP_REVISION \
-      org.opencontainers.image.licenses="GPL-3.0" \
-      org.opencontainers.image.base.name="docker.io/library/eclipse-temurin:25-jre-alpine"
-
```
Comment (balazs-szucs) [inline Dockerfile:56]: I think labels should probably stay. The versioning inferred via the .jar is good, but the Labels also contain other information that would be useful to be displayed on Dockerhub etc... and this is how it picks those up.

Comment (imnotjames) [reply Dockerfile:56]: The labels are injected by the docker metadata action.  You can see them in the screenshot I have of `docker image inspect`

Comment (balazs-szucs) [reply Dockerfile:56]: Indeed! My bad.

Thanks! :)
